DESIGN.md

```diff
@@ -0,0 +1,164 @@
+* installed under `NUPM_CURRENT_OVERLAY/lib/`
+
+You can also install non-Nushell packages as well using a "custom" project type where you specify a `build.nu` installation script
+(e.g., you can install Nushell itself with it).
+Plugins should also be supported, preferably not requiring fully custom `build.nu`.
+
+## Separate virtual environments
+
+Inspiration: Python, [conda](https://docs.conda.io/en/latest), `cargo`
+
+There are two different concepts in how to handle virtual environments:
+* Having global virtual environments, Python-style. We have a working prototype at https://github.com/kubouch/nuun using overlays.
+  * Installing a package will install it to the environment
+  * Possible to switch between them, they are completely isolated from each other
+* Per-project virtual environment, cargo-style
+  * A project has its own universe (like Rust projects, for example)
+
+Related to that is a lock file: It is intended to describe exactly the dependencies for a package so that it can be reproduced somewhere else.
+
+The overlays could be used to achieve all three goals at the same time. When installing a dependency for a package
+* `nupm` adds entry to a **lock file** (this should be the only file you need to 100% replicate the environment)
+* A .nu file (module) is auto-generated from the lock file and contains export statements like `export module NUPM_HOME/cache/packages/spam-v16.4.0-124ptnpbf/spam`. Calling `overlay use` on the file will activate your virtual environment, now you have a per-project environment
+* This file can be installed into a global location that's in your `NU_LIB_DIRS` (e.g., `NUPM_HOME/overlays`) -- now you have a global Python-like virtual environment
+
+Each package would basically have its own overlay. This overlay file (it's just a module) could be used to also handle dependencies. If your project depends on `foo` and `bar` which both depend on `spam` but different versions, they could both import the different verions privately in their own overlay files and in your project's overlay file would be just `export use path/to/foo` and `export use path/to/bar`. This should prevent name clashing of `spam`. The only problem that needs to be figured out is how to tell `foo` to be aware of its overlay.
+
+## Installation, bootstraping
+
+Requires these actions from the user (this should be kept as minimal as possible):
+* Add `NUPM_HOME/bin` to PATH (install location for binary projects)
+* Add `NUPM_HOME/lib` to NU_LIB_DIRS
+* Add `NUPM_HOME/overlays` to NU_LIB_DIRS
+* Make the `nupm` command available somehow (e.g., `use` inside `config.nu`)
+
+WIP: I have another idea in mind, need to think about it. The disadvantage of this is that the default install location is not an overlay.
+We could make `nupm` itself an overlay that adds itself as a command.
+
+There are several approaches:
+* bootstrap using shell script sourced from web (like rustup)
+* embedded inside Nushell's binary
```

Comment:
Embedding inside nushell binary would give us installation and upgrading for free. If anything it's the nushell itself that could use a bootstrap script